src/Database/LSMTree/Internal/IndexOrdinary.hs

```diff
@@ -0,0 +1,222 @@
+    fromSBS
+)
+where
+
+import Prelude hiding (length, drop, takeWhile)
+
+import Control.Monad
+       (
+           when
+       )
+import Data.Word
+       (
+           Word8,
+           Word16,
+           Word32,
+           Word64,
+           byteSwap32
+       )
+import Data.Primitive.ByteArray
+       (
+           ByteArray,
+           indexByteArray
+       )
+import Data.ByteString.Short
+       (
+           ShortByteString (ShortByteString)
+       )
+import Data.ByteString.Short qualified as ShortByteString
+       (
+           length
+       )
+import Data.Vector
+       (
+           Vector,
+           fromList,
+           length,
+           (!),
+           drop,
+           takeWhile,
+           findIndexR
+       )
+import Data.Vector.Primitive qualified as Primitive
+       (
+           Vector (Vector),
+           force,
+           null,
+           length,
+           take,
+           drop,
+           splitAt,
+       )
+import Database.LSMTree.Internal.Serialise
+       (
+           SerialisedKey (SerialisedKey')
+       )
+import Database.LSMTree.Internal.Entry
+       (
+           NumEntries (NumEntries)
+       )
+import Database.LSMTree.Internal.IndexCompact
+       (
+           PageNo (PageNo),
+           PageSpan (PageSpan)
+       )
+
+{-|
+    The version of the index and its serialisation format that is supported by
+    this module.
+-}
+supportedVersion :: Word32
+supportedVersion = 1
+
+{-|
+    A general-purpose fence pointer index.
+
+    An index is represented by a vector that maps the number of each page to the
+    key stored last in this page or, if the page is an overflow page, to the key
+    of the corresponding key–value pair. From this, the following constraints on
+    the vector follow:
+
+      * The vector must be non-empty.
+
+      * The elements of the vector must be non-decreasing.
+-}
+newtype IndexOrdinary = IndexOrdinary (Vector SerialisedKey)
+
+-- * Search
+
+{-|
+    Returns the span of pages that contain the key–value pair of a certain key
+    if there is such a pair, or an arbitrary, valid page span if there is not.
+-}
+search :: SerialisedKey -> IndexOrdinary -> PageSpan
+search key (IndexOrdinary lastKeys) = PageSpan (PageNo start) (PageNo end) where
+
+    start :: Int
+    start | protoStart < pageCount = protoStart
+          | otherwise              = maybe 0 succ $
+                                     findIndexR (/= lastKeys ! pred pageCount) $
+                                     lastKeys
+
+    end :: Int
+    end | protoStart < pageCount
+            = start + length (takeWhile (== key) (drop (succ start) lastKeys))
+        | otherwise
+            = pred pageCount
+
+    protoStart :: Int
+    protoStart = searchForProtoStart 0 (length lastKeys)
+
+    searchForProtoStart :: Int -> Int -> Int
+    searchForProtoStart first count
+        | count == 0               = first
+        | key <= lastKeys ! center = searchForProtoStart first frontCount
```

Comment:
You can use unsafe indexing here, as long as you are sure the number you're indexing at is not out of bounds

Reply:
It seems that it is guaranteed that the index is not out of bounds. Still, I’m reluctant to use unsafe indexing here. If I was wrong and the index *can* be out of bounds, the behavior of the software becomes unpredictable, to the point that we cannot rely on property-based tests anymore. The index checking is probably very cheap, and for each search there is only a logarithmic number of checks to be made (for a run comprising 4 GB it would be about 20 checks). I would favor safety over performance here. What is your opinion on this?

Reply:
Same reasoning as here https://github.com/IntersectMBO/lsm-tree/pull/244#discussion_r1639708395, if we locally use unsafe functions in a safe manner, then I think that's preferable

Reply:
In any case, this is now a topic for #256.